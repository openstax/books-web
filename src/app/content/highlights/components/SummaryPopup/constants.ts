export const mobilePaddingSides = 1.6;
export const mobilePaddingTopBottom = 1.4;
export const mobileMarginSides = 0.8;
export const mobileMarginTopBottom = 2;
