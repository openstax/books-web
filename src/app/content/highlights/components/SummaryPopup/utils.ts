import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const createHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
  const data = { target: JSON.stringify({ id, type: 'highlight' }) };
  const target = queryString.stringify(data) + `#${elementId}`;
  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`;
};
