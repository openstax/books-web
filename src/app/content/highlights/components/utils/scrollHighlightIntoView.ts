import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, element?: HTMLElement | null) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (element) {
    otherElements.push(element);
  }
  scrollIntoView(firstElement, otherElements);
};
