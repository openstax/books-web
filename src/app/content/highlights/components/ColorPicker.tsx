import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components/macro';
import { match, not } from '../../../fpUtils';
import { cardPadding, highlightStyles } from '../constants';
import ColorIndicator from './ColorIndicator';

interface SingleSelectProps {
  color?: string;
  onRemove: () => void;
  onChange: (color: string) => void;
  multiple: false | undefined;
}

interface MultipleSelectProps {
  selected: HighlightColorEnum[];
  onChange: (selected: string[]) => void;
  multiple: true;
}

type Props = (SingleSelectProps | MultipleSelectProps) & {
  className?: string;
  size?: 'small';
};

interface ColorButtonProps {
  name: string;
  checked: boolean;
  size: Props['size'];
  style: typeof highlightStyles[number];
  onClick: () => void;
  className?: string;
}

// tslint:disable-next-line:variable-name
const ColorButton = styled(({className, size, style, ...props}: ColorButtonProps) =>
  <FormattedMessage id={`i18n:highlighting:colors:${style.label}`}>
    {(msg: Element | string) =>
      <ColorIndicator
        style={style}
        size={size}
        title={msg}
        component={<label />}
        className={className}
      >
        <input type='checkbox' {...props} />
      </ColorIndicator>
    }
  </FormattedMessage>
)`
  cursor: pointer;
  margin: 0 ${cardPadding}rem ${cardPadding}rem 0;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
  }
`;

// tslint:disable-next-line:variable-name
const ColorPicker = ({className, ...props}: Props) => {

  return <div className={className}>
    {highlightStyles.map((style) => <ColorButton key={style.label}
      name={style.label}
      checked={props.multiple ? props.selected.includes(style.label) : props.color === style.label}
      style={style}
      onChange={() => props.multiple
        ? props.selected.includes(style.label)
          ? props.onChange(props.selected.filter(not(match(style.label))))
          : props.onChange([...props.selected, style.label])
        : props.color === style.label
          ? props.onRemove()
          : props.onChange(style.label)}
    />)}
  </div>;
};

export default styled(ColorPicker)`
  display: flex;
  flex-direction: row;
`;
