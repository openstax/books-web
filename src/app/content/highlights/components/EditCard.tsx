import { Highlight } from '@openstax/highlighter';
import { HighlightUpdateColorEnum } from '@openstax/highlighter/dist/api';
import { HighlightColorEnum } from '@openstax/highlighter/highlights-client/dist/models/Highlight';
import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import Button, { ButtonGroup } from '../../../components/Button';
import theme from '../../../theme';
import { assertWindow, mergeRefs } from '../../../utils';
import { clearFocusedHighlight, editStateChange, updateHighlight } from '../actions';
import { cardPadding, highlightStyles } from '../constants';
import { HighlightData } from '../types';
import ColorPicker from './ColorPicker';
import Confirmation from './Confirmation';
import Note from './Note';
import onClickOutside from './utils/onClickOutside';

interface Props {
  authenticated: boolean;
  loginLink: string;
  isFocused: boolean;
  hasUnsavedHighlight: boolean;
  highlight: Highlight;
  locationFilterId: string;
  pageId: string;
  onCreate: () => void;
  onBlur: typeof clearFocusedHighlight;
  onEditStateChange: typeof editStateChange;
  onSave: typeof updateHighlight;
  onRemove: () => void;
  onCancel: () => void;
  data?: HighlightData;
  className: string;
}

// tslint:disable-next-line:variable-name
const EditCard = React.forwardRef<HTMLElement, Props>((
  {
    authenticated,
    className,
    data,
    highlight,
    locationFilterId,
    pageId,
    isFocused,
    hasUnsavedHighlight,
    loginLink,
    onBlur,
    onCancel,
    onCreate,
    onEditStateChange,
    onRemove,
    onSave,
  }: Props,
  ref
) => {
  const defaultAnnotation = () => data && data.annotation ? data.annotation : '';
  const [pendingAnnotation, setPendingAnnotation] = React.useState<string>(defaultAnnotation());
  const [editingAnnotation, setEditing] = React.useState<boolean>(false);
  const [confirmingDelete, setConfirmingDelete] = React.useState<boolean>(false);
  const element = React.useRef<HTMLElement>(null);

  const trackCreateNote = useAnalyticsEvent('createNote');
  const trackEditNoteColor = useAnalyticsEvent('editNoteColor');
  const trackEditAnnotation = useAnalyticsEvent('editAnnotation');
  const trackDeleteHighlight = useAnalyticsEvent('deleteHighlight');

  const blurIfNotEditing = () => {
    if (!editingAnnotation) {
      onBlur();
    }
  };
 
  React.useEffect(() => {
    if (!isFocused && hasUnsavedHighlight) {
        setPendingAnnotation(defaultAnnotation());
        setEditing(false);
    }
  }, [isFocused, hasUnsavedHighlight]);

  React.useEffect(onClickOutside(element, isFocused, blurIfNotEditing), [isFocused, editingAnnotation]);

  const onColorChange = (color: HighlightColorEnum, isDefault?: boolean) => {
    highlight.setStyle(color);
    if (data) {
      onSave({
        highlight: {
          annotation: data.annotation,
          color: color as string as HighlightUpdateColorEnum,
        },
        id: data.id,
      }, {
        locationFilterId,
        pageId,
      });
      trackEditNoteColor(color);
    } else {
      assertWindow().getSelection().removeAllRanges();
      onCreate();
      trackCreateNote(isDefault ? 'default' : color);
    }
  };

  const saveAnnotation = (toSave: HighlightData) => {
    const addedNote = (data && data.annotation === undefined) ? true : false;

    onSave({
      highlight: {
        annotation: pendingAnnotation,
        color: toSave.color as string as HighlightUpdateColorEnum,
      },
      id: toSave.id,
    }, {
      locationFilterId,
      pageId,
    });
    trackEditAnnotation(addedNote, toSave.color);
    onCancel();
  };

  const updateUnsavedHighlightStatus = (newValue: string) => {
    if (!data) {
      return;
    }

    if (data.annotation !== newValue && !hasUnsavedHighlight) {
      onEditStateChange(true);
    }

    if ((data.annotation === newValue || (data.annotation === undefined && newValue === '')) && hasUnsavedHighlight) {
      onEditStateChange(false);
    }
  };

  const cancelEditing = () => {
    setPendingAnnotation(defaultAnnotation());
    setEditing(false);
    onEditStateChange(false);
    onCancel();
  };

  return <form
    className={className}
    ref={mergeRefs(ref, element)}
    data-analytics-region='edit-note'
  >
    <ColorPicker color={data ? data.color : undefined} onChange={onColorChange} onRemove={() => {
      if (data && !data.annotation && !pendingAnnotation) {
        onRemove();
        trackDeleteHighlight(data.color);
      }
    }} />
    <Note
      note={pendingAnnotation}
      onFocus={() => {
        if (!highlight.getStyle()) {
          onColorChange(highlightStyles[0].label, true);
        }
      }}
      onChange={(newValue) => {
        setPendingAnnotation(newValue);
        updateUnsavedHighlightStatus(newValue);
        setEditing(true);
      }}
    />
    {editingAnnotation && data && <ButtonGroup>
      <FormattedMessage id='i18n:highlighting:button:save'>
        {(msg: Element | string) => <Button
          data-testid='save'
          data-analytics-label='save'
          size='small'
          variant='primary'
          onClick={(e: React.FormEvent) => {
            e.preventDefault();
            setEditing(false);

            if (pendingAnnotation === '' && data.annotation) {
              setConfirmingDelete(true);
            } else {
              saveAnnotation(data);
            }
          }}
        >{msg}</Button>}
      </FormattedMessage>
      <FormattedMessage id='i18n:highlighting:button:cancel'>
        {(msg: Element | string) => <Button
          size='small'
          data-analytics-label='cancel'
          data-testid='cancel'
          onClick={(e: React.FormEvent) => {
            e.preventDefault();
            cancelEditing();
          }}
        >{msg}</Button>}
      </FormattedMessage>
    </ButtonGroup>}
    {confirmingDelete && data && <Confirmation
      data-testid='confirm-delete'
      data-analytics-region='highlighting-delete-note'
      message='i18n:highlighting:confirmation:delete-note'
      confirmMessage='i18n:highlighting:button:delete'
      onConfirm={() => saveAnnotation(data)}
      onCancel={() => {
        setEditing(true);
        setPendingAnnotation(defaultAnnotation());
      }}
      always={() => setConfirmingDelete(false)}
    />}
    {!authenticated && <Confirmation
      data-analytics-label='login'
      data-analytics-region='highlighting-login'
      message='i18n:highlighting:login:prompt'
      confirmMessage='i18n:highlighting:login:link'
      confirmLink={loginLink}
      onCancel={onBlur}
    />}
  </form>;
});

export default styled(EditCard)`
  background: ${theme.color.neutral.formBackground};
  user-select: none;
  overflow: visible;

  ${ButtonGroup} {
    margin-top: ${cardPadding}rem;
  }

  ${theme.breakpoints.mobile(css`
    visibility: hidden;
  `)}
`;
