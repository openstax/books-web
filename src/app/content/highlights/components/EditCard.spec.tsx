import { Highlight } from '@openstax/highlighter';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import createMockHighlight from '../../../../test/mocks/highlight';
import { testAccountsUser } from '../../../../test/mocks/userLoader';
import { makeFindByTestId } from '../../../../test/reactutils';
import * as selectAuth from '../../../auth/selectors';
import { formatUser } from '../../../auth/utils';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { assertDocument } from '../../../utils';
import { updateHighlight } from '../actions';
import { highlightStyles } from '../constants';
import ColorPicker from './ColorPicker';
import EditCard, { EditCardProps } from './EditCard';
import Note from './Note';
import * as onClickOutsideModule from './utils/onClickOutside';

jest.mock('./ColorPicker', () => (props: any) => <div mock-color-picker {...props} />);
jest.mock('./Note', () => (props: any) => <div mock-note {...props} />);
jest.mock('./Confirmation', () => (props: any) => <div mock-confirmation {...props} />);

describe('EditCard', () => {
  const highlight = createMockHighlight('asdf');
  const highlightData = highlight.serialize().data;
  const services = createTestServices();
  const store = createTestStore();
  const dispatch = jest.spyOn(store, 'dispatch');
  let editCardProps: Partial<EditCardProps>;

  beforeEach(() => {
    jest.resetAllMocks();
    highlight.elements = [assertDocument().createElement('span')];
    editCardProps = {
      highlight: highlight as unknown as Highlight,
      onBlur: jest.fn(),
      onCancel: jest.fn(),
      onCreate: jest.fn(),
      onRemove: jest.fn(),
      setAnnotationChangesPending: jest.fn(),
    };
  });

  it('matches snapshot when focused', () => {
    const data = {
      color: highlightStyles[0].label,
      ...highlightData,
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} data={data} isFocused={true} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with data', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} isFocused={true} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when editing', () => {
    highlight.getStyle.mockReturnValue('red');
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={highlightData}
              isFocused={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot without data', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} isFocused={true} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('chains ColorPicker onRemove', () => {
    const data = {
      ...highlightData,
      annotation: '',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              isFocused={true}
              data={data}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(editCardProps.onRemove).toHaveBeenCalled();
  });

  it('doesn\'t chain ColorPicker onRemove if there is a note', () => {
    const data = {
      ...highlightData,
      annotation: 'asdf',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} data={data} isFocused={true} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(editCardProps.onRemove).not.toHaveBeenCalled();
  });

  it('doesn\'t chain ColorPicker onRemove if there is a pending note', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: '',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={data}
              isFocused={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(editCardProps.onRemove).not.toHaveBeenCalled();
  });

  it('cancelling resets the form state', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={data}
              isFocused={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    expect(component.root.findAllByType('button').length).toBe(2);
    expect(note.props.note).toBe('asdf');

    const cancel = findByTestId('cancel');
    renderer.act(() => {
      cancel.props.onClick({preventDefault: jest.fn()});
    });

    expect(note.props.note).toBe('qwer');
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(component.root.findAllByType('button').length).toBe(0);
  });

  it('save saves', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={highlightData}
              locationFilterId='locationId'
              pageId='pageId'
              isFocused={true}
              onCreate={jest.fn()}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    expect(dispatch).toHaveBeenCalledWith(updateHighlight({
      highlight: {color: highlightData.style as any, annotation: 'asdf'},
      id: highlightData.id,
    }, {
      locationFilterId: 'locationId',
      pageId: 'pageId',
    }));
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(component.root.findAllByType('button').length).toBe(0);
    expect(editCardProps.onCancel).toHaveBeenCalled();
  });

  it('removing note shows confirmation', () => {
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={data}
              isFocused={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    expect(() => findByTestId('confirm-delete')).not.toThrow();
  });

  it('confirmation can save', () => {
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              locationFilterId='locationId'
              pageId='pageId'
              isFocused={true}
              data={data}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    const confirmation = findByTestId('confirm-delete');
    renderer.act(() => {
      confirmation.props.onConfirm();
      confirmation.props.always();
    });

    expect(() => findByTestId('confirm-delete')).toThrow();
    expect(dispatch).toHaveBeenCalledWith(updateHighlight({
      highlight: {color: highlightData.style as any, annotation: ''},
      id: highlightData.id,
    }, {
      locationFilterId: 'locationId',
      pageId: 'pageId',
    }));
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(editCardProps.onCancel).toHaveBeenCalled();
  });

  it('confirmation can cancel', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={data}
              isFocused={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    const confirmation = findByTestId('confirm-delete');
    renderer.act(() => {
      confirmation.props.onCancel();
      confirmation.props.always();
    });

    expect(() => findByTestId('confirm-delete')).toThrow();
    expect(dispatch).not.toHaveBeenCalled();
    expect(note.props.note).toBe('qwer');
  });

  it('responds to changes', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={data}
              isFocused={true}
              hasUnsavedHighlight={false}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const note = component.root.findByType(Note);

    renderer.act(() => {
      note.props.onChange('');
    });
    expect(editCardProps.setAnnotationChangesPending).toHaveBeenCalledWith(true);
  });

  it('dispatches if changes are reverted', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={data}
              isFocused={true}
              hasUnsavedHighlight={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );
    const note = component.root.findByType(Note);

    renderer.act(() => {
      note.props.onChange('qwer');
    });
    expect(editCardProps.setAnnotationChangesPending).toHaveBeenCalledWith(false);
  });

  it('handles color change when there is data', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              data={highlightData}
              locationFilterId='locationId'
              pageId='pageId'
              isFocused={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onChange('blue');
    });

    expect(highlight.setStyle).toHaveBeenCalledWith('blue');
    expect(store.dispatch).toHaveBeenCalledWith(updateHighlight({
      highlight: {annotation: highlightData.annotation, color: 'blue' as any},
      id: highlightData.id,
    }, {
      locationFilterId: 'locationId',
      pageId: 'pageId',
    }));
  });

  it('creates when changing color on a new highlight', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} isFocused={true} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onChange('blue');
    });

    expect(highlight.setStyle).toHaveBeenCalledWith('blue');
    expect(editCardProps.onCreate).toHaveBeenCalled();
  });

  it('sets color and creates when you focus', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onFocus();
    });

    expect(highlight.setStyle).toHaveBeenCalledWith(highlightStyles[0].label);
    expect(editCardProps.onCreate).toHaveBeenCalled();
  });

  it('focusing an existing note does nothing', () => {
    highlight.getStyle.mockReturnValue('red');
    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} data={highlightData} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onFocus();
    });

    expect(highlight.setStyle).not.toHaveBeenCalled();
    expect(editCardProps.onCreate).not.toHaveBeenCalled();
  });

  it('blurs when clicking outside', () => {
    const onClickOutside = jest.spyOn(onClickOutsideModule, 'default');
    onClickOutside.mockReturnValue(() => () => null);

    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard {...editCardProps} isFocused={true} />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    onClickOutside.mock.calls[0][2]({} as any);

    expect(component).toBeTruthy();
    expect(onClickOutside.mock.calls.length).toBe(1);
    expect(editCardProps.onBlur).toHaveBeenCalled();
  });

  it('doesn\'t blur when clicking outside and editing', () => {
    highlight.getStyle.mockReturnValue('red');

    const onClickOutside = jest.spyOn(onClickOutsideModule, 'default');
    onClickOutside.mockReturnValue(() => () => null);

    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              isFocused={true}
              data={highlightData}
              hasUnsavedHighlight={true}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    renderer.act(() => undefined);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    onClickOutside.mock.calls[1][2]({} as any);

    expect(onClickOutside.mock.calls.length).toBe(2);
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(editCardProps.onCancel).not.toHaveBeenCalled();
  });

  it('trackShowCreate for authenticated user', () => {
    const onClickOutside = jest.spyOn(onClickOutsideModule, 'default');
    onClickOutside.mockReturnValue(() => () => null);

    const mockSpyUser = jest.spyOn(selectAuth, 'user')
      .mockReturnValue(formatUser(testAccountsUser));

    const spyAnalytics = jest.spyOn(services.analytics.showCreate, 'track');

    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              isFocused={true}
              data={undefined}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>
    );

    expect(spyAnalytics).not.toHaveBeenCalled();

    // Wait for React.useEffect
    renderer.act(() => undefined);

    expect(() => component.root.findByProps({
      'data-analytics-region': 'highlighting-login',
    })).toThrow();
    expect(spyAnalytics).toHaveBeenCalled();
    mockSpyUser.mockClear();
  });

  it('call onHeightChange when element mounts', () => {
    const onClickOutside = jest.spyOn(onClickOutsideModule, 'default');
    onClickOutside.mockReturnValue(() => () => null);

    const onHeightChange = jest.fn();
    const createNodeMock = () => assertDocument().createElement('div');

    const component = renderer.create(
      <Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider onError={() => null}>
            <EditCard
              {...editCardProps}
              isFocused={true}
              onHeightChange={onHeightChange}
            />
          </MessageProvider>
        </Services.Provider>
      </Provider>,
      {createNodeMock}
    );

    expect(onHeightChange).not.toHaveBeenCalled();

    // Wait for mount
    renderer.act(() => undefined);

    expect(() => component.root.findByProps({
      'data-analytics-region': 'edit-note',
    })).not.toThrow();
    expect(onHeightChange).toHaveBeenCalled();
  });
});
