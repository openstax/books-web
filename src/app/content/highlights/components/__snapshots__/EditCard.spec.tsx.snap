// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCard matches snapshot when editing 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Neue Helvetica W01;
  border-radius: 0.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  font-size: 1.4rem;
  height: 3rem;
  padding: 0 2rem;
  color: #fff;
  background-color: #f36b32;
  font-weight: bold;
  border: none;
}

.c3:hover {
  background-color: #e96128;
}

.c3:active {
  background-color: #df571e;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Neue Helvetica W01;
  border-radius: 0.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  font-size: 1.4rem;
  height: 3rem;
  padding: 0 2rem;
  color: #5e6062;
  background-color: #fff;
  border: 1px solid #d5d5d5;
  font-weight: regular;
}

.c4:hover {
  background-color: #fafafa;
}

.c4:active {
  background-color: #e5e5e5;
}

.c2 {
  display: grid;
  overflow: visible;
  grid-auto-flow: column;
  grid-gap: 1rem;
}

.c0 {
  background: #f5f5f5;
  overflow: visible;
}

.c0 .c1 {
  margin-top: 0.8rem;
}

@media screen and (max-width:75em) {
  .c0 {
    visibility: hidden;
  }
}

<form
  className="c0"
  data-analytics-region="edit-note"
  onClick={[Function]}
>
  <div
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note="asdf"
    onChange={[Function]}
  />
  <div
    className="c1 c2"
  >
    <button
      className="c3"
      data-analytics-label="save"
      data-testid="save"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      className="c4"
      data-analytics-label="cancel"
      data-testid="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
  <div
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
  />
</form>
`;

exports[`EditCard matches snapshot when focused 1`] = `
.c0 {
  background: #f5f5f5;
  overflow: visible;
}

.c0 .c1 {
  margin-top: 0.8rem;
}

@media screen and (max-width:75em) {
  .c0 {
    visibility: hidden;
  }
}

<form
  className="c0"
  data-analytics-region="edit-note"
  onClick={[Function]}
>
  <div
    color="yellow"
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note=""
    onChange={[Function]}
  />
  <div
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
  />
</form>
`;

exports[`EditCard matches snapshot with data 1`] = `
.c0 {
  background: #f5f5f5;
  overflow: visible;
}

.c0 .c1 {
  margin-top: 0.8rem;
}

@media screen and (max-width:75em) {
  .c0 {
    visibility: hidden;
  }
}

<form
  className="c0"
  data-analytics-region="edit-note"
  onClick={[Function]}
>
  <div
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note=""
    onChange={[Function]}
  />
  <div
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
  />
</form>
`;

exports[`EditCard matches snapshot without data 1`] = `
.c0 {
  background: #f5f5f5;
  overflow: visible;
}

.c0 .c1 {
  margin-top: 0.8rem;
}

@media screen and (max-width:75em) {
  .c0 {
    visibility: hidden;
  }
}

<form
  className="c0"
  data-analytics-region="edit-note"
  onClick={[Function]}
>
  <div
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note=""
    onChange={[Function]}
  />
  <div
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
  />
</form>
`;
